src/arch/powerpc64.rs

```diff
@@ -0,0 +1,699 @@
+//! pointer near the stack base and updating it every time we switch into the
+//! the child (the coroutine) context using `switch_and_link`.
+//!
+//! For unwinding and backtraces to work as expected (that is, to continue in
+//! the parent after unwinding past the initial function of a child context),
+//! we need to use special DWARF CFI instructions to tell the unwinder how to
+//! find the parent frame.
+//!
+//! If you're curious a decent introduction to CFI things and unwinding is at
+//! <https://www.imperialviolet.org/2017/01/18/cfi.html>.
+//!
+//! ## Frame pointers
+//!
+//! Some tools or OSes do not use DWARF for stack unwinding, prefering to use
+//! the older (but simpler) frame pointer chain to capture a backtrace. This is
+//! particularly common in performance profiling tools such as Linux's perf
+//! callgraph profiler. These work by following a linked list of frame records
+//! starting from the FP register. Each record consists of 2 words: a pointer
+//! to the previous frame (aka the previous FP value) and the return address
+//! for this frame (aka the saved LR value).
+//!
+//! To support these tools, we also generate a valid stack frame record when
+//! switching into a coroutine. This works by treating the parent link at the
+//! root of the stack as a frame record which points to the top of the parent
+//! stack. The top of the parent stack contains the saved FP and LR values in
+//! the correct format for a frame record, which allows unwinding to continue on
+//! the parent stack.
+//!
+//! The LR value associated with the parent link is invalid since it points to
+//! the start of the initial function, but this shouldn't block the unwinding
+//! process.
+//!
+//! ## Stack layout
+//!
+//! Note: Non-volatile registers R29, R30, and R31 need to be explicitly
+//! saved/restored on our stack because they are used internally by LLVM and
+//! cannot be used as operands for inline asm (e.g., `inlate()`).
+//!
+//! Here is what the layout of the stack looks like when a coroutine is
+//! suspended.
+//!
+//! ```text
+//!
+//! +--------------+  <- Stack base
+//! | Initial func |  <- Only used once when resuming for the first time.
+//! +--------------+
+//! | Parent link  |  <- The Yielder is a pointer to this address. When the
+//! +--------------+     coroutine is running, it points to the top of the
+//! |              |     parent stack which contains a saved R29, R30, R31,
+//! ~    ...       ~     and LR just like a suspened coroutine.
+//! |              |
+//! +--------------+
+//! |  Saved R31   |
+//! +--------------+
+//! |  Saved R30   |
+//! +--------------+
+//! |  Saved R29   |
+//! +--------------+
+//! |  Saved LR    |
+//! +--------------+
+//! |  Link area   |  <- 32-byte link area
+//! +--------------+
+//! ```
+//!
+//! And this is the layout of the parent stack when a coroutine is running:
+//!
+//! ```text
+//! +--------------+
+//! |     ...      |
+//! |  saved LR    |  <- 32-byte link area.
```

Comment:
The saved LR goes into the link area at the bottom the top here is just `...` to indicate whatever was on the stack before.